### ISSUE TITLE
BLD: issues building on Ubuntu22.04 (C++17/C++14 related?)

### ISSUE THREAD

**Issue by andyfaff** [scipy/scipy#21786]
### Describe your issue.

Having issues building (locally) on Ubuntu 22.04, looking to update some of the CI for GPU testing.
This is done in a fresh docker container. The there's a whole bunch of error messages, but they all seem to be related to c++17/c++14 parameter passing.

g++ versio

### Reproducing Code Example

```python
apt update
cd src
apt-get install -y --no-install-recommends python3-dev libopenblas-dev pkg-config gfortran build-essential python3.10-venv
python3 -m venv ../venvs/scipy-venv
source ../venvs/scipy-venv/bin/activate
python -m pip install build
python -m build . --wheel
g++ --version
```

### Error message

```shell
[78/1274] Compiling C++ object scipy/special/_ufuncs.cpython-310-aarch64-linux-gnu.so.p/xsf_wrappers.cpp.o
  In file included from ../scipy/special/xsf/cephes/kolmogorov.h:60,
                   from ../scipy/special/xsf/stats.h:9,
                   from ../scipy/special/xsf_wrappers.cpp:26:
  ../scipy/special/xsf/cephes/dd_real.h: In function ‘std::pair<xsf::cephes::detail::double_double, xsf::cephes::detail::double_double> xsf::cephes::detail::divrem(const xsf::cephes::detail::double_double&, const xsf::cephes::detail::double_double&)’:
  ../scipy/special/xsf/cephes/dd_real.h:386:107: note: parameter passing for argument of type ‘std::pair<xsf::cephes::detail::double_double, xsf::cephes::detail::double_double>’ when C++17 is enabled changed to match C++14 in GCC 10.1
    386 |                                                                                   const double_double &rhs) {
        |                                                                                                           ^
```

### SciPy/NumPy/Python version and system information

```shell
main
(scipy-venv) root@ac402ec5f830:/src# g++ --version
g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by tylerjereddy** [comment]
I'm on the same Ubuntu version (it is what I use for releases as well), and when I force usage of the same version of the GNU toolchain similar steps work fine for me. I'm on Python 3.11 and outside a docker container though.

**Comment by rgommers** [comment]
That's a little odd, since we do have a CI job with GCC 9 that is passing. 

Is the issue localized to `special/xsf/`, or happening in other places as well?

**Comment by andyfaff** [comment]
it also happens in `_lib/boost_math` as well. I've put a short build log at https://gist.github.com/andyfaff/bd71068d10515a37f27c153928e1f00c#file-short_log-txt and the full build log at https://gist.github.com/andyfaff/bd71068d10515a37f27c153928e1f00c#file-full_log-txt.

In the full log I'm seeing:

```
Using cached compile:
Cached command line:  c++ /src/build/meson-private/tmprvra6fq3/testfile.cpp -o /src/build/meson-private/tmprvra6fq3/output.obj -c -D_FILE_OFFSET_BITS=64 -O0 -fpermissive 

Code:
 
        #include <version>
        int main(void) {
            /* If it's not defined as a macro, try to use as a symbol */
            #ifndef _LIBCPP_VERSION
                _LIBCPP_VERSION;
            #endif
            return 0;
        }
Cached compiler stdout:
 
Cached compiler stderr:
 /src/build/meson-private/tmprvra6fq3/testfile.cpp: In function 'int main()':
/src/build/meson-private/tmprvra6fq3/testfile.cpp:6:17: error: '_LIBCPP_VERSION' was not declared in this scope
    6 |                 _LIBCPP_VERSION;
      |                 ^~~~~~~~~~~~~~~

```

**Comment by rgommers** [comment]
These logs show a quite different build config than you noted in the issue description, namely GCC 13.2.0 on Linux aarch64. That's a combo we don't test in CI; we only have GCC 11.4.0 on Ubuntu 22.04 (see [example Cirrus CI log](https://cirrus-ci.com/task/5603426648719360)). So it could well have an issue that we didn't encounter before.

Can you fix the issue description if it's really about gcc 13.2.0 on Linux aarch64?

**Comment by andyfaff** [comment]
Sorry, in order to get the logs I took a shortcut which used a more recent version of gcc+Ubuntu.

The issue does exist with gcc11.4.0 on Ubuntu 22.04 (I am using aarch64). I redid in a 22.04 container, see the log [here](https://gist.github.com/andyfaff/bd71068d10515a37f27c153928e1f00c#file-redux_2204-txt). You'll note that this was from trying to do `python -m build --wheel .`.

The meson log file from a `python dev.py build` is [here](https://gist.github.com/andyfaff/bd71068d10515a37f27c153928e1f00c#file-redux-mesonlog-txt).

**Comment by andyfaff** [comment]
When I try and change the platform (`podman run -it -v $PWD:/src --arch=x86_64 ubuntu:22.04`) the build also fails. log [here](https://gist.githubusercontent.com/andyfaff/bd71068d10515a37f27c153928e1f00c/raw/a5504267e4440b77cd3f11827e5974007e282d4a/x86_64.txt).

The fail is slightly different.

Here the output from `python -m build . --wheel` is probably more useful;
```
[75/1274] Compiling C++ object scipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o
FAILED: scipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o 
c++ -Iscipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p -Iscipy/special -I../scipy/special -Iscipy/_lib -I../scipy/_lib -I../scipy/_build_utils/src -I../../tmp/build-env-j2fpk0md/lib/python3.10/site-packages/numpy/_core/include -I/usr/include/python3.10 -I/usr/include/x86_64-linux-gnu/python3.10 -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O3 -fPIC -DSP_SPECFUN_ERROR -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -Wno-psabi -MD -MQ scipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o -MF scipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o.d -o scipy/special/_special_ufuncs.cpython-310-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o -c ../scipy/special/_special_ufuncs.cpp
c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
[80/1274] Compiling C++ object scipy/special/_ufuncs_cxx.cpython-310-x86_64-linux-gnu.so.p/meson-generated__ufuncs_cxx.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by rgommers** [comment]
I just re-ran the Linux aarch64 wheels build ([Cirrus CI log](https://cirrus-ci.com/build/5663880863547392)) to check if it was a regression - but those all pass.

If this used to work, perhaps you can bisect it to find the problem?

**Comment by andyfaff** [comment]
Thanks for the git bisect reminder @rgommers. I think I've fixed my problem, and it wasn't a scipy issue. I've been using (brew installed) podman as a replacement for docker.

I started off by trying to git bisect in the podman, but couldn't even build really old tags. I then reinstalled docker and could do a successful HEAD build in ostensibly the same container, with the same commands (note - I still see the build noise outlined in the original post).

I then did a search for `c++: fatal error: Killed signal terminated program cc1plus` and there were several posts indicating that this could happen if the build process was using too many resources (e.g. out of memory, too much parallelisation).

This lead me to think that the podman machine didn't have enough resources by default, but docker probably does. I gave the podman machine more resources by:

```
podman machine stop
podman machine set --cpus 8 --memory 8192
podman machine start
podman run -it -v $PWD:/src ubuntu:22.04
```

Following that reconfiguration the build works. I suspect it was mainly a memory problem. By default the podman machine is set to 2048 MB.

Thus, the moral of the story is, if you're trying to build in a container using podman/docker, the container needs to be given enough resources.

Feel free to close. I didn't close to give people a chance to read the post.
